File: src/backend/app/schemas/workspace.py
"""
Workspace and StrategyUnit Pydantic schemas.
"""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.trading import TradingGatewayConfig, TradingSnapshot

# ---------------------------------------------------------------------------
# Workspace schemas
# ---------------------------------------------------------------------------


class WorkspaceCreate(BaseModel):
    """Create workspace request."""

    name: str = Field(..., max_length=200, description="Workspace name")
    description: str | None = Field(None, max_length=500, description="Description")
    workspace_type: str = Field("research", description="Workspace type: research/trading")
    settings: dict[str, Any] = Field(default_factory=dict, description="Settings JSON")
    trading_config: dict[str, Any] = Field(
        default_factory=dict, description="Trading settings JSON"
    )


class WorkspaceUpdate(BaseModel):
    """Update workspace request."""

    name: str | None = Field(None, max_length=200)
    description: str | None = Field(None, max_length=500)
    workspace_type: str | None = None
    settings: dict[str, Any] | None = None
    trading_config: dict[str, Any] | None = None


class WorkspaceResponse(BaseModel):
    """Workspace response."""

    id: str
    user_id: str
    name: str
    description: str | None = None
    workspace_type: str = "research"
    settings: dict[str, Any] = Field(default_factory=dict)
    trading_config: dict[str, Any] = Field(default_factory=dict)
    unit_count: int = 0
    completed_count: int = 0
    status: str = "idle"
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class WorkspaceListResponse(BaseModel):
    """Workspace list response."""

    total: int
    items: list[WorkspaceResponse]


# ---------------------------------------------------------------------------
# StrategyUnit schemas
# ---------------------------------------------------------------------------


class StrategyUnitCreate(BaseModel):
    """Create strategy unit request."""

    group_name: str = Field("", max_length=200, description="Group name")
    strategy_id: str | None = Field(None, description="Strategy template ID")
    strategy_name: str = Field("", max_length=200, description="Strategy display name")
    symbol: str = Field("", max_length=50, description="Trading symbol code")
    symbol_name: str = Field("", max_length=200, description="Symbol display name")
    timeframe: str = Field("1d", max_length=10, description="K-line timeframe")
    timeframe_n: int = Field(1, ge=1, description="Timeframe multiplier")
    category: str = Field("", max_length=100, description="Classification tag")
    data_config: dict[str, Any] = Field(default_factory=dict)
    unit_settings: dict[str, Any] = Field(default_factory=dict)
    params: dict[str, Any] = Field(default_factory=dict)
    optimization_config: dict[str, Any] = Field(default_factory=dict)
    trading_mode: str = Field("paper", description="Trading mode: paper/live")
    gateway_config: TradingGatewayConfig | dict[str, Any] = Field(default_factory=dict)
    lock_trading: bool = False
    lock_running: bool = False
    trading_snapshot: TradingSnapshot | dict[str, Any] = Field(default_factory=dict)


class StrategyUnitBatchCreate(BaseModel):
    """Batch create strategy units request."""

    units: list[StrategyUnitCreate] = Field(..., min_length=1)


class StrategyUnitUpdate(BaseModel):
    """Update strategy unit request."""

    group_name: str | None = None
    strategy_id: str | None = None
    strategy_name: str | None = None
    symbol: str | None = None
    symbol_name: str | None = None
    timeframe: str | None = None
    timeframe_n: int | None = None
    category: str | None = None
    data_config: dict[str, Any] | None = None
    unit_settings: dict[str, Any] | None = None
    params: dict[str, Any] | None = None
    optimization_config: dict[str, Any] | None = None
    trading_mode: str | None = None
    gateway_config: TradingGatewayConfig | dict[str, Any] | None = None
    lock_trading: bool | None = None
    lock_running: bool | None = None
    trading_instance_id: str | None = None
    trading_snapshot: TradingSnapshot | dict[str, Any] | None = None


class StrategyUnitResponse(BaseModel):
    """Strategy unit response."""

    id: str
    workspace_id: str
    group_name: str = ""
    strategy_id: str | None = None
    strategy_name: str = ""
    symbol: str = ""
    symbol_name: str = ""
    timeframe: str = "1d"
    timeframe_n: int = 1
    category: str = ""
    sort_order: int = 0
    data_config: dict[str, Any] = Field(default_factory=dict)
    unit_settings: dict[str, Any] = Field(default_factory=dict)
    params: dict[str, Any] = Field(default_factory=dict)
    optimization_config: dict[str, Any] = Field(default_factory=dict)
    trading_mode: str = "paper"
    gateway_config: dict[str, Any] = Field(default_factory=dict)
    lock_trading: bool = False
    lock_running: bool = False
    trading_instance_id: str | None = None
    trading_snapshot: dict[str, Any] = Field(default_factory=dict)
    run_status: str = "idle"
    run_count: int = 0
    last_run_time: float | None = None
    last_task_id: str | None = None
    last_optimization_task_id: str | None = None
    bar_count: int | None = None
    metrics_snapshot: dict[str, Any] = Field(default_factory=dict)
    opt_status: str | None = None
    opt_total: int | None = None
    opt_completed: int | None = None
    opt_progress: float | None = None
    opt_elapsed_time: float | None = None
    opt_remaining_time: float | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class StrategyUnitListResponse(BaseModel):
    """Strategy unit list response."""

    total: int
    items: list[StrategyUnitResponse]


class UnitRuntimeFileResponse(BaseModel):
    """Runtime file metadata exposed to the workspace UI."""

    name: str
    relative_path: str
    size: int
    kind: str


class UnitRuntimeInfoResponse(BaseModel):
    """Runtime directory metadata for a strategy unit."""

    unit_id: str
    runtime_dir: str
    log_dir: str | None = None
    files: list[UnitRuntimeFileResponse] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Bulk operation schemas
# ---------------------------------------------------------------------------


class BulkDeleteRequest(BaseModel):
    """Bulk delete request."""

    ids: list[str] = Field(..., min_length=1)


class SortRequest(BaseModel):
    """Reorder units request."""

    unit_ids: list[str] = Field(..., min_length=1, description="Ordered list of unit IDs")


class GroupRenameRequest(BaseModel):
    """Group rename request."""

    unit_ids: list[str] = Field(..., min_length=1)
    mode: str = Field(
        "custom",
        description="Rename mode: custom/strategy/symbol/symbol_name/category/replace",
    )
    value: str = Field("", description="New name for custom mode")
    search: str = Field("", description="Search string for replace mode")
    replace: str = Field("", description="Replace string for replace mode")


class UnitRenameRequest(BaseModel):
    """Unit rename request (single unit display name override)."""

    unit_id: str
    mode: str = Field("custom")
    value: str = Field("")
    search: str = Field("")
    replace: str = Field("")


# ---------------------------------------------------------------------------
# Run orchestration schemas
# ---------------------------------------------------------------------------


class RunUnitsRequest(BaseModel):
    """Run selected strategy units."""

    unit_ids: list[str] = Field(..., min_length=1)
    parallel: bool = Field(False, description="Run in parallel if True, sequential if False")


class StopUnitsRequest(BaseModel):
    """Stop selected running strategy units."""

    unit_ids: list[str] = Field(..., min_length=1)


class UnitStatusResponse(BaseModel):
    """Unit run status response for polling."""

    id: str
    run_status: str
    last_task_id: str | None = None
    metrics_snapshot: dict[str, Any] = Field(default_factory=dict)
    run_count: int = 0
    last_run_time: float | None = None
    bar_count: int | None = None
    trading_instance_id: str | None = None
    trading_snapshot: dict[str, Any] = Field(default_factory=dict)
    trading_mode: str = "paper"
    lock_trading: bool = False
    lock_running: bool = False
    # Optimization progress (filled during polling)
    opt_status: str | None = None
    opt_total: int | None = None
    opt_completed: int | None = None
    opt_progress: float | None = None
    opt_elapsed_time: float | None = None
    opt_remaining_time: float | None = None


# ---------------------------------------------------------------------------
# Optimization schemas (Phase 4)
# ---------------------------------------------------------------------------


class ParamRangeSpec(BaseModel):
    """Parameter range for optimization."""

    start: float
    end: float
    step: float
    type: str = "float"


class UnitOptimizationRequest(BaseModel):
    """Submit optimization for a strategy unit."""

    unit_id: str
    param_ranges: dict[str, ParamRangeSpec]
    n_workers: int = Field(default=4, ge=1, le=32)
    mode: str = Field("grid", description="Optimization mode: grid/random")
    timeout: int = Field(0, ge=0, description="Timeout in seconds (0 = no limit)")


class ApplyBestParamsRequest(BaseModel):
    """Apply best parameters from optimization to a unit."""

    unit_id: str
    optimization_task_id: str
    result_index: int = Field(0, ge=0, description="Index of result to apply (0 = best)")


class OptimizationArtifactResponse(BaseModel):
    """Persisted optimization artifact metadata."""

    artifact_path: str | None = None
    artifact_manifest_path: str | None = None
    artifact_summary_path: str | None = None
    artifact_status: str | None = None
    artifact_error: str | None = None
    optimization_task_id: str
    optimization_result_index: int
    trial_index: int | None = None


# ---------------------------------------------------------------------------
# Report schemas (Phase 5)
# ---------------------------------------------------------------------------


class ReportCreateRequest(BaseModel):
    """Create / recalculate combined report with config."""

    start_date: str | None = Field(None, description="Statistics start date (ISO)")
    end_date: str | None = Field(None, description="Statistics end date (ISO)")
    max_cash: float | None = Field(None, description="Override max invested cash")
    calc_method: str = Field("simple", description="Calculation method: simple/compound")
    annual_days: int = Field(252, ge=1, description="Trading days per year")
    weight_mode: str = Field("equal", description="Weight mode: equal/custom")
    weights: dict[str, float] = Field(
        default_factory=dict, description="Unit ID -> weight (custom mode)"
    )


class ReportDeleteRequest(BaseModel):
    """Delete report (clear cached report for workspace)."""

    pass

File: src/frontend/src/types/workspace.ts
/**
 * Workspace and StrategyUnit TypeScript types.
 * Iteration 124 - Strategy Research feature.
 */

export type UnitRunStatus = 'idle' | 'queued' | 'running' | 'completed' | 'failed' | 'cancelled' | 'timeout'
export type WorkspaceStatus = 'idle' | 'running' | 'completed' | 'error'
export type ViewMode = 'card' | 'table'
export type WorkspaceDataSourceType = 'csv' | 'mysql' | 'postgresql' | 'mongodb'
export type WorkspaceType = 'research' | 'trading'
export type TradingMode = 'paper' | 'live'
export type ReportCalcMethod = 'simple' | 'compound'
export type ReportWeightMode = 'equal' | 'custom'
export type OptimizationDisplayViewMode = 'table' | 'analysis'
export type OptimizationSortDirection = 'asc' | 'desc'

export interface WorkspaceDataSourceConfig {
  type: WorkspaceDataSourceType
  csv: {
    directory_path: string
    delimiter: string
    encoding: string
    has_header: boolean
  }
  mysql: {
    host: string
    port: number
    database: string
    username: string
    password: string
    table: string
  }
  postgresql: {
    host: string
    port: number
    database: string
    schema: string
    username: string
    password: string
    table: string
  }
  mongodb: {
    uri: string
    database: string
    collection: string
    username: string
    password: string
    auth_source: string
  }
}

export interface WorkspaceReportConfig {
  start_date?: string | null
  end_date?: string | null
  stat_range?: [string | null, string | null]
  max_cash?: number | null
  calc_method?: ReportCalcMethod | string
  annual_days?: number
  weight_mode?: ReportWeightMode | string
  weights?: Record<string, number>
  visible_fields?: string[]
}

export interface WorkspaceOptimizationDisplayConfig {
  sort_key?: string
  sort_dir?: OptimizationSortDirection
  view_mode?: OptimizationDisplayViewMode
  calc_method?: ReportCalcMethod | string
  annual_days?: number
  stat_time_range?: [string | null, string | null]
  visible_fields?: string[]
  analysis_params?: string[]
  analysis_metric?: string
}

export interface WorkspaceSettings extends Record<string, unknown> {
  data_source?: WorkspaceDataSourceConfig
  report_config?: WorkspaceReportConfig
  optimization_config?: WorkspaceOptimizationDisplayConfig
}

export interface GatewayConfig {
  preset_id?: string | null
  name?: string | null
  params?: Record<string, unknown>
}

export interface TradingPosition {
  data_name: string
  direction: string
  size: number
  price: number | null
  current_price: number | null
  market_value: number | null
  margin_value?: number | null
  multiplier?: number | null
  margin_rate?: number | null
  commission?: number | null
  gross_pnl?: number | null
  pnl: number | null
  pnlcomm?: number | null
  position_pnl?: number | null
  updated_at?: string | null
  data_time?: string | null
  source?: string | null
  position_source?: string | null
  asset_spec_source?: string | null
  valuation_status?: string | null
  valuation_warnings?: string[]
}

export interface TradingTrade {
  id: string
  data_name: string
  direction: string
  size: number
  price: number | null
  value: number | null
  commission: number | null
  pnl: number | null
  pnlcomm: number | null
  barlen: number | null
  dtopen: string | null
  dtclose: string | null
  datetime: string | null
}

export interface TradingSnapshot {
  instance_id: string | null
  instance_status: string
  mode: TradingMode
  error: string | null
  started_at: string | null
  stopped_at: string | null
  gateway_summary: string | null
  long_position: number
  short_position: number
  today_pnl: number | null
  position_pnl: number | null
  latest_price: number | null
  change_pct: number | null
  long_market_value: number | null
  short_market_value: number | null
  leverage: number | null
  cumulative_pnl: number | null
  max_drawdown_rate: number | null
  trading_day: string | null
  updated_at: string | null
  detail_route: string | null
  position_source?: string | null
  asset_spec_source?: string | null
  valuation_status?: string | null
  valuation_warnings?: string[]
  open_order_cancel?: Record<string, unknown> | null
  positions: TradingPosition[]
  trades: TradingTrade[]
}

export interface TradingAutoSession {
  name: string
  open: string
  close: string
}

export interface TradingAutoConfig {
  enabled: boolean
  buffer_minutes: number
  sessions: TradingAutoSession[]
  scope: string
}

export interface TradingAutoScheduleItem {
  session: string
  start: string
  stop: string
  market_open?: string | null
  market_close?: string | null
}

export interface TradingPositionManagerItem {
  unit_id: string
  unit_name: string
  symbol: string
  symbol_name: string | null
  trading_mode: TradingMode
  long_position: number
  short_position: number
  avg_price: number | null
  latest_price: number | null
  position_pnl: number
  market_value: number
  margin_value?: number | null
  multiplier?: number | null
  margin_rate?: number | null
  commission?: number | null
  gross_pnl?: number | null
  updated_at?: string | null
  data_time?: string | null
  data_name?: string | null
  position_source?: string | null
  asset_spec_source?: string | null
  valuation_status?: string | null
  valuation_warnings?: string[]
}

export interface TradingPositionManagerResponse {
  positions: TradingPositionManagerItem[]
  total_long_value: number
  total_short_value: number
  total_pnl: number
}

export interface TradingDailySummaryItem {
  trading_date: string
  daily_pnl: number
  trade_count: number
  cumulative_pnl: number
  max_drawdown: number
}

export interface TradingDailySummaryResponse {
  summaries: TradingDailySummaryItem[]
}

export interface StrategyUnitRuntimeFile {
  name: string
  relative_path: string
  size: number
  kind: string
}

export interface StrategyUnitRuntimeInfo {
  unit_id: string
  runtime_dir: string
  log_dir: string | null
  files: StrategyUnitRuntimeFile[]
}

// ---------------------------------------------------------------------------
// Workspace
// ---------------------------------------------------------------------------

export interface Workspace {
  id: string
  user_id: string
  name: string
  description: string | null
  workspace_type: WorkspaceType
  settings: WorkspaceSettings
  trading_config: Record<string, unknown>
  unit_count: number
  completed_count: number
  status: WorkspaceStatus
  created_at: string
  updated_at: string
}

export interface WorkspaceCreate {
  name: string
  description?: string
  workspace_type?: WorkspaceType
  settings?: WorkspaceSettings
  trading_config?: Record<string, unknown>
}

export interface WorkspaceUpdate {
  name?: string
  description?: string
  workspace_type?: WorkspaceType
  settings?: WorkspaceSettings
  trading_config?: Record<string, unknown>
}

export interface WorkspaceListResponse {
  total: number
  items: Workspace[]
}

// ---------------------------------------------------------------------------
// Strategy Unit
// ---------------------------------------------------------------------------

export interface StrategyUnit {
  id: string
  workspace_id: string
  group_name: string
  strategy_id: string | null
  strategy_name: string
  symbol: string
  symbol_name: string
  timeframe: string
  timeframe_n: number
  category: string
  sort_order: number
  data_config: Record<string, unknown>
  unit_settings: Record<string, unknown>
  params: Record<string, unknown>
  optimization_config: Record<string, unknown>
  trading_mode: TradingMode
  gateway_config: GatewayConfig
  lock_trading: boolean
  lock_running: boolean
  trading_instance_id: string | null
  trading_snapshot: TradingSnapshot
  run_status: UnitRunStatus
  run_count: number
  last_run_time: number | null
  last_task_id: string | null
  last_optimization_task_id: string | null
  bar_count: number | null
  metrics_snapshot: Record<string, unknown>
  // Optimization progress (populated by polling)
  opt_status?: string | null
  opt_total?: number | null
  opt_completed?: number | null
  opt_progress?: number | null
  opt_elapsed_time?: number | null
  opt_remaining_time?: number | null
  opt_started_at_ms?: number | null
  opt_last_sync_at_ms?: number | null
  created_at: string
  updated_at: string
}

export interface StrategyUnitCreate {
  group_name?: string
  strategy_id?: string
  strategy_name?: string
  symbol?: string
  symbol_name?: string
  timeframe?: string
  timeframe_n?: number
  category?: string
  data_config?: Record<string, unknown>
  unit_settings?: Record<string, unknown>
  params?: Record<string, unknown>
  optimization_config?: Record<string, unknown>
  trading_mode?: TradingMode
  gateway_config?: GatewayConfig
  lock_trading?: boolean
  lock_running?: boolean
  trading_snapshot?: Partial<TradingSnapshot>
}

export interface StrategyUnitUpdate {
  group_name?: string
  strategy_id?: string
  strategy_name?: string
  symbol?: string
  symbol_name?: string
  timeframe?: string
  timeframe_n?: number
  category?: string
  data_config?: Record<string, unknown>
  unit_settings?: Record<string, unknown>
  params?: Record<string, unknown>
  optimization_config?: Record<string, unknown>
  trading_mode?: TradingMode
  gateway_config?: GatewayConfig
  lock_trading?: boolean
  lock_running?: boolean
  trading_instance_id?: string | null
  trading_snapshot?: Partial<TradingSnapshot>
}

export interface StrategyUnitListResponse {
  total: number
  items: StrategyUnit[]
}

// ---------------------------------------------------------------------------
// Bulk operations
// ---------------------------------------------------------------------------

export interface BulkDeleteRequest {
  ids: string[]
}

export interface SortRequest {
  unit_ids: string[]
}

export interface GroupRenameRequest {
  unit_ids: string[]
  mode: 'custom' | 'strategy' | 'symbol' | 'symbol_name' | 'category' | 'replace'
  value?: string
  search?: string
  replace?: string
}

export interface UnitRenameRequest {
  unit_id: string
  mode: 'custom' | 'strategy' | 'symbol' | 'symbol_name' | 'category' | 'replace'
  value?: string
  search?: string
  replace?: string
}

// ---------------------------------------------------------------------------
// Run orchestration
// ---------------------------------------------------------------------------

export interface RunUnitsRequest {
  unit_ids: string[]
  parallel?: boolean
}

export interface StopUnitsRequest {
  unit_ids: string[]
}

export interface UnitStatusResponse {
  id: string
  run_status: UnitRunStatus
  last_task_id: string | null
  metrics_snapshot: Record<string, unknown>
  run_count: number
  last_run_time: number | null
  bar_count: number | null
  trading_instance_id: string | null
  trading_snapshot: TradingSnapshot
  trading_mode: TradingMode
  lock_trading: boolean
  lock_running: boolean
  opt_status: string | null
  opt_total: number | null
  opt_completed: number | null
  opt_progress: number | null
  opt_elapsed_time: number | null
  opt_remaining_time: number | null
}

export interface RunUnitResult {
  unit_id: string
  task_id: string | null
  status: string
  error?: string
}

// ---------------------------------------------------------------------------
// Optimization (Phase 4)
// ---------------------------------------------------------------------------

export interface ParamRangeSpec {
  start: number
  end: number
  step: number
  type?: string
}

export interface UnitOptimizationRequest {
  unit_id: string
  param_ranges: Record<string, ParamRangeSpec>
  n_workers?: number
  mode?: string
  timeout?: number
}

export interface ApplyBestParamsRequest {
  unit_id: string
  optimization_task_id: string
  result_index?: number
}

export interface OptimizationSubmitResult {
  task_id: string
  unit_id: string
  total_combinations: number
  n_workers: number
}

export interface OptimizationArtifactResponse {
  artifact_path: string | null
  artifact_manifest_path: string | null
  artifact_summary_path: string | null
  artifact_status: string | null
  artifact_error: string | null
  optimization_task_id: string
  optimization_result_index: number
  trial_index: number | null
}

export interface OptimizationProgressResponse {
  task_id?: string | null
  status: string
  total: number
  completed: number
  failed: number
  progress?: number | null
  elapsed_time?: number | null
  remaining_time?: number | null
  error?: string | null
}

export interface OptimizationResultRow {
  params?: Record<string, number | string>
  [key: string]: unknown
}

export interface OptimizationResultsResponse {
  task_id: string
  status: string
  total: number
  completed: number
  failed: number
  param_names: string[]
  objective?: string
  best?: OptimizationResultRow | null
  rows?: OptimizationResultRow[]
  results?: OptimizationResultRow[]
}

export interface CancelOptimizationResponse {
  task_id: string
  cancelled: boolean
}

export interface ApplyBestParamsResponse {
  unit_id: string
  applied_params: Record<string, unknown>
  metrics: Record<string, unknown>
}

export interface WorkspaceReportCreateRequest {
  start_date?: string | null
  end_date?: string | null
  max_cash?: number | null
  calc_method?: ReportCalcMethod
  annual_days?: number
  weight_mode?: ReportWeightMode
  weights?: Record<string, number>
}

export interface WorkspaceReportUnitReference {
  id: string
  strategy_name: string
  symbol: string
  symbol_name: string
  timeframe: string
  group_name: string
  category: string
  run_status: UnitRunStatus
  run_count: number
  last_run_time: number | null
  last_task_id: string | null
  start_date: string | null
  data_source: string
  value?: number | null
}

export interface WorkspaceReportUnitRow extends WorkspaceReportUnitReference {
  total_return?: number | null
  annual_return?: number | null
  sharpe_ratio?: number | null
  max_drawdown?: number | null
  win_rate?: number | null
  total_trades?: number | null
  profitable_trades?: number | null
  losing_trades?: number | null
  initial_cash?: number | null
  final_value?: number | null
  net_value?: number | null
  net_profit?: number | null
  max_leverage?: number | null
  max_market_value?: number | null
  max_drawdown_value?: number | null
  adjusted_return_risk?: number | null
  avg_profit?: number | null
  avg_profit_rate?: number | null
  total_win_amount?: number | null
  total_loss_amount?: number | null
  profit_loss_ratio?: number | null
  profit_factor?: number | null
  profit_rate_factor?: number | null
  profit_loss_rate_ratio?: number | null
  odds?: number | null
  daily_avg_return?: number | null
  daily_max_loss?: number | null
  daily_max_profit?: number | null
  weekly_avg_return?: number | null
  weekly_max_loss?: number | null
  weekly_max_profit?: number | null
  monthly_avg_return?: number | null
  monthly_max_loss?: number | null
  monthly_max_profit?: number | null
  trading_cost?: number | null
  trading_days?: number | null
  [key: string]: unknown
}

export interface WorkspaceReportSummary {
  total_units: number
  completed_units: number
  avg_total_return: number | null
  avg_annual_return: number | null
  avg_sharpe_ratio: number | null
  avg_max_drawdown: number | null
  avg_win_rate: number | null
  total_trades: number | null
  best_return_unit: WorkspaceReportUnitReference | null
  worst_drawdown_unit: WorkspaceReportUnitReference | null
  config: WorkspaceReportCreateRequest
}

export interface WorkspaceReportResponse {
  workspace_id: string
  workspace_name: string
  summary: WorkspaceReportSummary
  units: WorkspaceReportUnitRow[]
}

export interface WorkspaceReportDeleteResponse {
  workspace_id: string
  cleared: boolean
  message: string
}
